### ISSUE TITLE
Lightweight?

### ISSUE THREAD

**Issue by drtrigon** [f3d-app/f3d#759]
**Is your feature request related to a problem? Please describe.**
I recognized the new release (thanks for it!) has become 70MB bigger than before. Version 1.3.1 was without raytracing about 35MB and now it is about 100MB (with raytracing even 200MB).

**Describe the solution you'd like**
Would it be possible to have a "minimal" version, that removes not so often used features in favor of a smaller size?

Would it be possible to move more features in to optional plugins such that users like me can choose which to add? E.g. a plugin per file format or similar?

**Describe alternatives you've considered**
Is it possible to remove some of the libraries that it is packed with? If yes wicht ones?

**Additional context**
For me the main improvement of 2.0.0 is "Added an interaction (down arrow) to load all files from the directory containing the current file" and to me it seems a bit overkill to add an additional of 70MB just for that.

I do understand that this is a trade-off and e.g. features requested by myself like STEP support are to parts the reason why f3d grows in size. I think "big" the features should be broken up an distributed over various optional plugins.

**Contribution**
I am happy testing anything you want me to! ;)

**Comment by mwestphal** [comment]
The other main feature you seemed to have missed is the new packaging system,[ f3d-superbuild](https://github.com/f3d-app/f3d-superbuild).
Technically, it means that instead of creating a single lib containing everything, it is now split in many libraries, which implies a relatively large overhead by files.

The only big librairies in there is the raytracing stuff, and that is why we provide a non-raytracing version.
You can easily check in the package the size of the files in it.


```
   48.2 MiB [#############]  libOpenImageDenoise.so.1
   30.5 MiB [########     ]  libembree3.so.3
   15.9 MiB [####         ]  libvtkFiltersCore.so.1
   13.1 MiB [###          ]  libvtkCommonCore.so.1
    9.8 MiB [##           ]  libopenvkl_module_cpu_device_8.so.1
    9.2 MiB [##           ]  libopenvkl_module_cpu_device_4.so.1
    8.5 MiB [##           ]  libopenvkl_module_cpu_device_16.so.1
    7.3 MiB [#            ]  libvtkCommonDataModel.so.1
    7.1 MiB [#            ]  libvtkCommonMisc.so.1
    7.0 MiB [#            ]  libvtkFiltersGeneral.so.1
    6.7 MiB [#            ]  libospray_module_ispc.so.2
    5.1 MiB [#            ]  libTKGeomAlgo.so.7
    5.0 MiB [#            ]  libTKGeomBase.so.7
    4.6 MiB [#            ]  libvtkhdf5.so.1
    4.0 MiB [#            ]  libvtkRenderingVolume.so.1
    4.0 MiB [#            ]  libTKBool.so.7
    3.9 MiB [#            ]  libTKIGES.so.7
    3.8 MiB [#            ]  libvtkInteractionWidgets.so.1
    3.4 MiB [             ]  libTKSTEP.so.7
    3.0 MiB [             ]  libvtkRenderingOpenGL2.so.1
    3.0 MiB [             ]  libvtkImagingCore.so.1
    3.0 MiB [             ]  libTKTopAlgo.so.7
    2.8 MiB [             ]  libvtkRenderingCore.so.1
    2.7 MiB [             ]  libTKBO.so.7
    2.7 MiB [             ]  libTKXSBase.so.7
    2.6 MiB [             ]  libTKSTEPBase.so.7
    2.6 MiB [             ]  libTKMath.so.7
    2.6 MiB [             ]  libTKShHealing.so.7
    2.5 MiB [             ]  libassimp.so.5
    2.3 MiB [             ]  libvtkFiltersExtraction.so.1
    2.3 MiB [             ]  libvtkIOGeometry.so.1
    2.2 MiB [             ]  libAlembic.so.1.8
    2.0 MiB [             ]  libdraco.so.8
    2.0 MiB [             ]  libvtkIOImage.so.1
    1.9 MiB [             ]  libTKernel.so.7
    1.8 MiB [             ]  libTKSTEPAttr.so.7
    1.8 MiB [             ]  libvtkFiltersGeometry.so.1
    1.6 MiB [             ]  libvtkCommonExecutionModel.so.1
    1.5 MiB [             ]  libf3d.so
    1.4 MiB [             ]  libvtkFiltersModeling.so.1
    1.3 MiB [             ]  libvtkIOXML.so.1
    1.3 MiB [             ]  libvtkParallelDIY.so.1
    1.3 MiB [             ]  libvtkRenderingAnnotation.so.1
    1.3 MiB [             ]  libvtkFiltersStatistics.so.1
    1.2 MiB [             ]  libvtkFiltersParallel.so.1
    1.2 MiB [             ]  libTKBRep.so.7
    1.2 MiB [             ]  libTKG3d.so.7
    1.1 MiB [             ]  libvtkFiltersHybrid.so.1
    1.1 MiB [             ]  libvtknetcdf.so.1
    1.1 MiB [             ]  libvtkFiltersSources.so.1
    1.0 MiB [             ]  libTKMesh.so.7
    1.0 MiB [             ]  libvtkImagingGeneral.so.1
  940.0 KiB [             ]  libvtkglew.so.1
  856.0 KiB [             ]  libvtkIOExodus.so.1
  852.0 KiB [             ]  libvtkIOLegacy.so.1
  836.0 KiB [             ]  libvtkImagingHybrid.so.1

```

That being said, we could try to provide a statically built or a minimal version of F3D that focus on being as lightweigth as possible. I will give it a go when I can :).

Keep in mind you can also build F3D yourself for your system and your specific needs.

**Comment by Meakk** [comment]
> Is it possible to remove some of the libraries that it is packed with? If yes wicht ones?

If you downloaded the binaries, I think all the plugins are static (please correct me if I'm wrong @mwestphal), so you cannot remove the dependencies.

Out of curiosity, why is 100MB too much for you? That seems neglectable on a basic computer.

**Comment by mwestphal** [comment]
> If you downloaded the binaries, I think all the plugins are static (please correct me if I'm wrong @mwestphal), so you cannot remove the dependencies.

Correct, but we can turn off some plugins in a smaller binary package if needed.

**Comment by Meakk** [comment]
Ok, so maybe we should just build all plugins as dynamic plugins in the superbuild and let the user select the plugins he wants to install in NSIS (maybe there's something similar for .deb files?)

**Comment by mwestphal** [comment]
I'm afraid this is technically not possible.

**Comment by mwestphal** [comment]
So I gave it a quick go:

- txz full feature (raytracing): 87 Mb 
- txz without raytracing: 35 Mb 
- txz without any optional component: 21 Mb
- txz with statically built VTK without any optional component:  11 Mb

So there is two ways forward with this.

1. Provide a super small downloadable static version of F3D, simple but no customizable
2. Explore how we could rely on dynamic plugin in the binary package to let user download/install only what they need, more work but better in terms of fonctionnality

Let us know what you think @drtrigon

**Comment by drtrigon** [comment]
I acknowledge superbuild is supercool as it allows you to have such a variety of build options (e.g. portable which I like very much) but I prefer to use the binaries and not to go into the bussiness of building it myself - especially on win which I am forced to use at work... ;)

> That being said, we could try to provide a statically built or a minimal version of F3D that focus on being as lightweigth as possible. I will give it a go when I can :).

That would be awesome brilliant! Thanks for considering!

I run a portable system of a SD card as productive office system, here every byte counts... ;) Also I consider f3d as replacement for FreeCAD or Blender in order to view such files, at some point - if it becomes comparable in size - it would become kind of useless to me as then I would prefer to use one of the other since they also allow editing...

> Ok, so maybe we should just build all plugins as dynamic plugins in the superbuild and let the user select the plugins he wants to install in NSIS (maybe there's something similar for .deb files?

Would be a nice idea. As I love the portable version it would be nice to have a plugin folder in there where I am allowed to delete whatever looks useless or rarely used to me. E.g. raytracing; brilliant to know f3d can do it - if I need it - but for everyday work I won't need it at all. When I need it, I will download full f3d again. Another - more involved option - would be to have an online tool that allows you to select what to add before downloading (the portable) but I think that is lot of overhead on your side.

What is the unpacked size of the 35MB txz (w/o raytracing)... the current 2.0.0 w/o raytracing is the same size but 100MB if unpacked... 1.3.1 was 35MB unpacked.

So to me very useful would be:
  - a plugin folder in there where I am allowed to delete whatever looks useless or rarely used to me
  - or distribution of main exec and plugins separately such that I can put together what I need (or add later when I recogize taht I need it)

I think the first proposal is preferred as it is simpler and produces less overhead to you and all users that do not care.

**Comment by Meakk** [comment]
Got it.  
Since you mentioned STEP files earlier, please note that STEP file support won't be available in the minimal build @mwestphal is suggesting.

**Comment by drtrigon** [comment]
> Got it. Since you mentioned STEP files earlier, please note that STEP file support won't be available in the minimal build @mwestphal is suggesting.

STEP is important cause there are only a few apps supporting it. So this means the super small build is not a solution here. (even though it sounds intressting also)

**Comment by Meakk** [comment]
Here is what we decided offline:
- Build all plugins in dynamic mode so they will be located in `lib/`
- Build external dependencies in static mode so plugins becomes self-contained

So basically, you will be able to safely remove plugins you are not interested in. Also, plugins binaries should be smaller.

On top of that, we will provide a minimal package containing a single self-contained executable without any plugin (except the `native` one containing VTK readers) but it will not support STEP files.

**Comment by drtrigon** [comment]
This is perfect! That will allow me (and others) to put f3d togheter exactly as I need it. Thanks a lot to everybody involved!

For which release is it planned?

**Comment by Meakk** [comment]
Probably 2.1, there is no plan for a 2.0.1 yet.  
However, you'll be informed through this issue and you will be able to download a nightly build.

**Comment by drtrigon** [comment]
> We actually have gone another way after testing diffferent configuration. The provided binaries are not build statically, you will not be able to remove any plugin however building statically drastically reduce the size of the package.

I am a bit confused; is it statically linked or not? ;)

> Once untarred, the linux release is now just a bit over 100 Mb, 60 Mb less than before.
> 
> Please note, in these 100 Mb there is the STEP/IGES support that represent 40Mb, which you need afaiu.

I usually use the non-raytracing version, it is now ~45MB compared to ~35MB in v1.3.1. That is ok for me and a lot better than ~100MB. Thanks a lot for that!

Do I remeber right that before this issue was raised raytracing was by default included in v2.0.0? However never mind, I like having the choice and the size is ok for me now.

(Btw.: I am aware that I am part of the problem due to STEP support - I am always amazed to see that "simple" STEP supports involves a 40MB library binary... ;)

**Comment by drtrigon** [comment]
Addendum: using https://github.com/upx/upx I got this;

```
> upx.exe -v --lzma --backup *.*
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2023
UPX 4.0.2       Markus Oberhumer, Laszlo Molnar & John Reiser   Jan 30th 2023

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
upx: concrt140.dll: CantPackException: CFGuard enabled PE files are not supported (use --force to disable)
upx: dbghelp.dll: CantPackException: CFGuard enabled PE files are not supported (use --force to disable)
   3986432 ->   3295232   82.66%    win64/pe     draco.dll
  42312704 ->  11541504   27.28%    win64/pe     f3d.dll
  468480 ->    165888   35.41%    win64/pe     f3d.exe
  147968 ->     56320   38.06%    win64/pe     F3DShellExtension.dll
upx: msvcp140.dll: CantPackException: CFGuard enabled PE files are not supported (use --force to disable)
upx: msvcp140_1.dll: CantPackException: CFGuard enabled PE files are not supported (use --force to disable)
upx: msvcp140_2.dll: CantPackException: CFGuard enabled PE files are not supported (use --force to disable)
upx: msvcp140_atomic_wait.dll: CantPackException: CFGuard enabled PE files are not supported (use --force to disable)
upx: msvcp140_codecvt_ids.dll: CantPackException: CFGuard enabled PE files are not supported (use --force to disable)
    310272 ->     77312   24.92%    win64/pe     tbb12.dll
upx: vcruntime140.dll: CantPackException: CFGuard enabled PE files are not supported (use --force to disable)
upx: vcruntime140_1.dll: CantPackException: CFGuard enabled PE files are not supported (use --force to disable)
   --------------------   ------   -----------   -----------
  47225856 ->  15136256   32.05%                 [ 5 files ]

Packed 14 files: 5 ok, 9 errors.
```
UPX was able to compress relevant files down to less than a third of the size. This was done using fairly conservative option and low compression.

May be it would be worth adding UPX compression to your build chain, as least as an additional release? What do you think?

**Comment by mwestphal** [comment]
We rely on cpack to create the package and CPack does not seem to support upx:
https://cmake.org/cmake/help/latest/cpack_gen/archive.html

Looks like there is an issue for it (more or less) without activity for 4 years : https://gitlab.kitware.com/cmake/cmake/-/issues/18122

It may be possible to implement a custom logic to use UPX  (eg see this: https://codereview.stackexchange.com/questions/103972/cmake-module-to-make-executable-as-small-as-possible) but it feel not necessary and out of scope of F3D project to handle such complexity.

So I'd say, not until it is natively supported by CPack.

**Comment by Meakk** [comment]
Before considering this kind of tool, we could think of a `MinSizeRel` build type.

**Comment by drtrigon** [comment]
Too bad. I am amazed by the results of UPX, LESS THAN 33% of the size! And it would only be adding one single line to the build script, wouldn't it? However I can do it myself when I download a new release. ;)

From my understanding I don't think `MinSizeRel` is comparable, because it optimizes for size instead of speed (trades on for the other). The concept of UPX is different: Increase the speed by decresing the size (and by theat the amount of instructions needed to execute in order to achive the same result). A quite appealing concept IMHO. :)

**Comment by mwestphal** [comment]
> And it would only be adding one single line to the build script, wouldn't it? However I can do it myself when I download a new release. ;)

Not at all, as there is no "build script". You are welcome to try though.
